src/display.jl

```diff
@@ -114,9 +117,24 @@ function display_dict(x)
     end
 
     return data
-
 end
 
+display_dict(::Nothing) = nothing
+
+"""
+Generate a `(data=..., metadata=...)` tuple.
+
+This defaults to using `display_dict` and `metadata` respectively (see the
+documentaiton for those functions to see what `data` and `metadata` should be).
+A custom method for this function can be specified whenever it is difficult to
+compute the data and metadata separately. Most users will just need to define a
+method for `display_dict`.
+
+This function may also return `nothing`, in which case no Jupyter message is
+sent at all.
+"""
+display_payload(x) = (data=display_dict(x), metadata=metadata(x))
```

Comment:
This is fine with me in principle, but I'd prefer to separate it from IJulia somehow, so that modules that want to support this don't need to pull in an IJulia dependency.

Maybe we can define it as a `repr` method with a custom MIME type?  e.g. you define
```jl
Base.repr(::MIME"application/x-ijulia-payload", x::MyObject) = (data=display_dict(x), metadata=metadata(x))
```

Reply:
Yeah absolutely I see what you mean. I'm not hugely a fan of hacking `repr` with a custom MIME (everything else to my knowledge returns a string which I think is more in-line with how MIMEs are meant to work), but without a suitable alternative, I think it's... fine. Alternatively we could just tell them to use Requires.jl (which is what things like WebIO use already).

Also just another good use case for better integrated optional dependencies :^). It seems like Julia's package ecosystem isn't great right now for package relations that arent a very strong "A depends on B".

Reply:
Better to create a small package that both IJulia and other packages could depend on.

Reply:
MIMEBundle.jl?

Reply:
Sure.

Reply:
Probably `MIMEBundles.jl` for the package and module, and `MIMEBundle` for a type declared in the module.

Reply:
Where should this package live? I can volunteer the Gizmos org (a lot of that org is already concerned with rich content so it makes sense) but wherever is fine with me.

Reply:
Probably the same org as IJulia, i.e. JuliaLang.   Or maybe IJulia should be moved to JuliaGizmos?

Reply:
I guess we can put it in JuliaGizmos

Reply:
I don't feel super strongly, I just think that we should put it in a place that makes sense. If you'd like me to create the package in JuliaGizmos, I can do that and add some starter code.

Reply:
Sure, please gobahead

Reply:
https://github.com/JuliaGizmos/MIMEBundles.jl

Wrote some initial stuff. I also took an opportunity to clean-up some API (but we should be able to forward the current IJulia functions to the equivalent for MIMEBundles).